### PR TITLE
persist: Add `FixedSizeBytesStats`

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds a new variant of `BytesStats` called `FixedSizedBytesStats`. It's very similar to `AtomicBytesStats` in that they can't be trimmed, but they're explicitly for types that implement the new `FixedSizeCodec`.

### Motivation

Related to https://github.com/MaterializeInc/materialize/issues/24830

As we evolve our encodings in Persist, what bytes are recorded will change. For example, today stats for `ScalarType::Time` are `AtomicBytesStats` of a `ProtoNaiveTime`, but with our new columnar encodings it will be a `PackedNaiveTime`. We need a way to distinguish between the two.

An alternative is adding a "kind" field to `AtomicBytesStats`, I originally implemented this but we determined the better approach would be a whole separate type which allows for a cleaner separation.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/repr/src/row/encoding.rs [hunk 1]
```diff
@@ -414,7 +419,11 @@ impl DatumToPersist for NullableProtoDatumToPersist {
             let col = col.downcast_ref::<Option<Vec<u8>>>()?;
             debug_assert!(validity.is_superset(col.logical_nulls().as_ref()));
             let (lower, upper, null_count) = proto_datum_min_max_nulls(col, ValidityRef::none());
-            let stats = BytesStats::Atomic(AtomicBytesStats { lower, upper });
+            let stats = BytesStats::Atomic(AtomicBytesStats {
+                lower,
+                upper,
+                kind: Some(AtomicBytesStatsKind::ProtoDatum),
```
Comment (bkirwi) [inline src/repr/src/row/encoding.rs:425]: At first glance this looks like a [compatibility hazard](https://materializeinc.slack.com/archives/C03P1EGB7A7/p1719326449752639); thoughts? Anyways worth running nightlies probably...

**Review by bkirwi — COMMENTED** [review]
IIRC we discussed using a part-level version number to manage migrations at the actual data level; wondering if that would make sense here.

Personally I kind of appreciate the `PackedBytesStats` idea... while in many ways it is similar to the atomic stats as you mention, there are important differences: it has a meaningful sort, and it can be computed off the columnar representation instead of the raw data. (One option would be to call it `FixedLengthBytesStats` or whatever, which stresses that while the proto types are similar we're dealing with a different sort of thing at the arrow level.)

**Comment by ParkMyCar** [comment]
> IIRC we discussed using a part-level version number to manage migrations at the actual data level; wondering if that would make sense here.

For sure! I thought about adding a version number to stats, and it definitely should be possible, just a bit heavier weight. In the meantime I realized at the moment the only type of stats that we really need to evolve are the `AtomicBytesStats` and just adding a "kind" here was faster than adding a version.

> Personally I kind of appreciate the `PackedBytesStats` idea... while in many ways it is similar to the atomic stats as you mention, there are important differences: it has a meaningful sort, and it can be computed off the columnar representation instead of the raw data. (One option would be to call it `FixedLengthBytesStats` or whatever, which stresses that while the proto types are similar we're dealing with a different sort of thing at the arrow level.)

Totally fair, and it is a cleaner separation between the older stats types and the newer types. I'll update this PR to add a `PackedBytesStats` instead of adding a "kind" to `AtomicBytesStats`

File: src/repr/src/row/encoding.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/repr/src/row/encoding.rs:425]: Eeek yeah you're right, let me update this!

**Review by bkirwi — APPROVED** [review]
🙏

**Pull request merged**